Python/chapter01/1.1 - Is Unique/miguel_1.1_sol.py

```diff
@@ -0,0 +1,101 @@
+"""
+Python version 3.7.0
+1.1
+Implement an algorithm to determine if a string has all unique characters.
+What if you cannot use additional data structures?
+What if the input string is empty?
+I assume that the user will input a non-empty string
+If the string is empty, then raise a ValueError exception
+If the input is NOT a string, then raise a TypeError exception
+"""
+import unittest
+
+
+def check_is_unique_input(input_str):
+    """
+    Does some input checking for the is_unique function and its variants
+    Raises exceptions if input_str is empty or if input_str is not of type string
+    :param input_str: a string that we are checking to make
+    :return:
+    """
+    if not isinstance(input_str, str):
+        raise TypeError("expected string as input")
+    if input_str == "":
+        raise ValueError("empty input string")
+
+
+def is_unique(input_str):
+    """
+    Determines if input_str has all unique characters.
+    input string MUST have at least 1 character. n>=1, where n is the number of characters in a string.
+    Hence, the domain is now defined. Any input not within the domain will not be considered. (n < 1)
+    An exception will be raised to handle faulty input.
```

Comment:
it's a good practice to define what "faulty input" means in the docstring for your public interface. This is formally known as [preconditions](https://en.wikipedia.org/wiki/Precondition).